### ISSUE TITLE
Return `Tuple` type containing `Record` type has an error

### ISSUE THREAD

**Issue by dlee992** [numba/numba#9199]
```python
import os

import numpy as np
from numba import njit, types

os.environ["NUMBA_DEBUG_TYPEINFER"] = "1"
os.environ["NUMBA_CAPTURED_ERRORS"] = "new_style"

arr = np.array(object=[(1, 2)], dtype=[('a1', 'f8'), ('a2', 'f8')])

@njit
def _goo(arr, i):
    rec = arr[i]
    return rec, i

@njit
def foo(arr, i):
    rec, j = _goo(arr=arr, i=i)

foo(arr=arr, i=0)
```
run this, get an error 
```python
  File "numba/core/callwrapper.py", line 175, in build_wrapper
    status, retval = self.context.call_conv.call_function(
  File "numba/core/callconv.py", line 600, in call_function
    out = self.context.get_returned_value(builder, resty, retval)
  File "numba/core/base.py", line 674, in get_returned_value
    return self.data_model_manager[ty].from_return(builder, val)
  File "numba/core/datamodel/models.py", line 627, in from_return
    return self._from("from_data", builder, vals)
  File "numba/core/datamodel/models.py", line 561, in _from
    v = getattr(dm, methname)(builder, val)
  File "numba/core/datamodel/models.py", line 967, in from_data
    raise NotImplementedError("use load_from_data_pointer() instead")
NotImplementedError: ('use load_from_data_pointer() instead' ...
```

If I only return rec in _goo, then it's fine.

**Comment by dlee992** [comment]
I dig into this issue, turned out that I think we just need a conversion from `[8 * i8]` to its pointer type, then everything should be fine. How to do this using `llvmlite.ir`? I tried to do this, but still have a little problem to this.

**Comment by gmarkall** [comment]
Thanks for the report! I can reproduce this issue. A smaller test also seems to reproduce the 

```
NotImplementedError: ('use load_from_data_pointer() instead'
```

error message:

```python
from numba import njit, types

fields = [('a1', {'type': types.float64, 'offset': 0})]
ty = types.Record(fields, 8, True)


@njit((ty[::1],))
def foo(arr):
    return arr[0], 0
```

Note that if we fix this issue, I think the next issue we'll run into will be the issue with reference counting of records from record arrays - the underlying data might get deleted before it's returned: https://github.com/numba/numba/issues/1315 / https://github.com/numba/numba/issues/6947 / https://github.com/numba/numba/issues/6835 / https://github.com/numba/numba/issues/8051 (a list also in https://github.com/numba/numba/issues/8051#issuecomment-1126582823)

**Comment by dlee992** [comment]
Yeah, I think you nailed it. @gmarkall 
I tried with this fix
```python
    def from_data(self, builder, value):
        # alloca a Record ptr and make it point to value
        # but have an issue, after boxing the record back to python,
        # the contents in Record seems random
        ptr = cgutils.alloca_once_value(builder, value)
        builder.store(value, ptr)
        return ptr
        # raise NotImplementedError("use load_from_data_pointer() instead")
```

It runs fine in jitted func, but when boxing it back to python, the contents in the record seems random value.

**Comment by dlee992** [comment]
So you mean after a fix, we could meet use-after-free bugs? If we don't return it back to python, do you see any trouble we could meet? I think at least we won't have memory leak issue?

**Comment by gmarkall** [comment]
> So you mean after a fix, we could meet use-after-free bugs?

Yes - that would probably explain the random value of the record when it is boxed.

> If we don't return it back to python (I mean we only return record to another jitted function), do you see any trouble we could meet? I think at least we won't have memory leak issue?

NRT will also be prone to deallocating the underlying data, so we're not safe even if the record remains inside a jitted function.

**Comment by dlee992** [comment]
@gmarkall , Oh, then my fix above is just allocating a pointer and let it point to the Record value. But the recarray holding this value will be freed, since its refcount is down to 0. So when I try to use the pointer to get the value in python or other jitted func, it could make troubles, right?

**Comment by gmarkall** [comment]
From discussion in the dev meeting: probably a reasonable and lower effort way to address this issue for now is to fix Numba such that when this use of a Record is attempted, a more helpful error message explaining that use of a Record in this way is not supported, instead of the current error:

```
NotImplementedError: ('use load_from_data_pointer() instead'
```

**Comment by sklam** [comment]
I started a patch (that I don't have time to finish in the near-term) to add refcount tracking to `Record` type. Diff at https://github.com/numba/numba/compare/main...sklam:enh/refct_record?expand=1

The branch is minimally tested with this script:
```python
import os

import numpy as np
from numba import njit, types

# os.environ["NUMBA_DEBUG_TYPEINFER"] = "1"
os.environ["NUMBA_CAPTURED_ERRORS"] = "new_style"

dt = np.dtype([('a1', 'f8'), ('a2', 'f8')])

@njit
def _goo(i):
    arr = np.zeros(3, dtype=dt)
    rec = arr[i]
    return rec

@njit
def foo(i):
    rec = _goo(i=i)
    rec.a1 = 1.0
    rec.a2 = 3.0
    rec2 = _goo(i=i) # This will likely zero `rec` due to freelist reuse w/o proper refcounting
    return rec

rec = foo(i=0)
print(rec)
assert rec['a1'] == 1.0
assert rec['a2'] == 3.0
```

Many things still segfaults as many things are not fully connected yet.

**Comment by dlee992** [comment]
And for a general educational purpose, if we can give some examples to compare how to use llvmlite to implement some basic C operations, I would be grateful. E.g., in C, we can allocate a pointer, and point it to a value, and do some assignments or so, and we can also see how to do the same thing using llvmlite. Or maybe we already have these materials buried in somewhere?

**Comment by gmarkall** [comment]
The Python version of the LLVM Kaleidoscope tutorial is a bit old now (I don't think it's maintained) but it might be a good source of examples: https://github.com/eliben/pykaleidoscope/

**Comment by gmarkall** [comment]
cc @yxdragon for updates as #10004 reports a similar issue.

**Comment by yxdragon** [comment]
@gmarkall so, is there a way to create a union dtype? like this:
```python
# return a, b
return np.zeros((a, b), dtype=np.dtype([('first', a.dtype), ('second', b.dtype)]))
```
but np.dtype is not supported.
